// @antv/g6/esm/registry/extension/index.js
import { getExtension } from '..';
import { arrayDiff } from '../../utils/diff';
import { parseExtensions } from '../../utils/extension';
export class ExtensionController {
    context;
    extensions = [];
    extensionMap = {};
    constructor(context) {
        this.context = context;
    }
    setExtensions(extensions) {
        const stdExtensions = parseExtensions(this.category, extensions);
        const { enter, update, exit, keep } = arrayDiff(this.extensions, stdExtensions, (extension) => extension.key);
        this.createExtensions(enter);
        this.updateExtensions([...update, ...keep]);
        this.destroyExtensions(exit);
        this.extensions = stdExtensions;
    }
    createExtension(extension) {
        const { category } = this;
        const { key, type } = extension;
        const Ctor = getExtension(category, type);
        if (!Ctor)
            return;
        const instance = new Ctor(this.context, extension);
        this.extensionMap[key] = instance;
    }
    createExtensions(extensions) {
        extensions.forEach((extension) => this.createExtension(extension));
    }
    updateExtension(extension) {
        const { key } = extension;
        const instance = this.extensionMap[key];
        if (instance) {
            instance.update(extension);
        }
    }
    updateExtensions(extensions) {
        extensions.forEach((extension) => this.updateExtension(extension));
    }
    destroyExtension(key) {
        const instance = this.extensionMap[key];
        if (instance) {
            instance.destroy();
            delete this.extensionMap[key];
        }
    }
    destroyExtensions(extensions) {
        extensions.forEach(({ key }) => this.destroyExtension(key));
    }
    destroy() {
        Object.values(this.extensionMap).forEach((extension) => extension.destroy());
        // @ts-expect-error force delete
        this.context = {};
        this.extensions = [];
        this.extensionMap = {};
    }
}
/**
 * <zh/> 模块实例基类
 *
 * <en/> Base class for extension instance
 */
export class BaseExtension {
    context;
    options;
    events = [];
    destroyed = false;
    constructor(context, options) {
        this.context = context;
        this.options = options;
    }
    update(options) {
        this.options = Object.assign(this.options, options);
    }
    destroy() {
        // @ts-expect-error force delete
        this.context = {};
        // @ts-expect-error force delete
        this.options = {};
        this.destroyed = true;
    }
}
