// @antv/g-canvas/dist/index.js
'use strict';

var tslib = require('tslib');
var gLite = require('@antv/g-lite');
var CanvasPathGenerator = require('@antv/g-plugin-canvas-path-generator');
var CanvasPicker = require('@antv/g-plugin-canvas-picker');
var CanvasRenderer = require('@antv/g-plugin-canvas-renderer');
var DomInteraction = require('@antv/g-plugin-dom-interaction');
var HTMLRenderer = require('@antv/g-plugin-html-renderer');
var ImageLoader = require('@antv/g-plugin-image-loader');
var util = require('@antv/util');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var CanvasPathGenerator__namespace = /*#__PURE__*/_interopNamespaceDefault(CanvasPathGenerator);
var CanvasPicker__namespace = /*#__PURE__*/_interopNamespaceDefault(CanvasPicker);
var CanvasRenderer__namespace = /*#__PURE__*/_interopNamespaceDefault(CanvasRenderer);
var DomInteraction__namespace = /*#__PURE__*/_interopNamespaceDefault(DomInteraction);
var HTMLRenderer__namespace = /*#__PURE__*/_interopNamespaceDefault(HTMLRenderer);
var ImageLoader__namespace = /*#__PURE__*/_interopNamespaceDefault(ImageLoader);

var Canvas2DContextService = /** @class */ (function () {
    function Canvas2DContextService(context) {
        this.renderingContext = context.renderingContext;
        this.canvasConfig = context.config;
    }
    Canvas2DContextService.prototype.init = function () {
        var _a = this.canvasConfig, container = _a.container, canvas = _a.canvas;
        if (canvas) {
            this.$canvas = canvas;
            if (container &&
                canvas.parentElement !== container) {
                container.appendChild(canvas);
            }
            this.$container = canvas.parentElement;
            this.canvasConfig.container = this.$container;
        }
        else if (container) {
            // create container
            this.$container = util.isString(container)
                ? document.getElementById(container)
                : container;
            if (this.$container) {
                // create canvas
                var $canvas = document.createElement('canvas');
                this.$container.appendChild($canvas);
                if (!this.$container.style.position) {
                    this.$container.style.position = 'relative';
                }
                this.$canvas = $canvas;
            }
        }
        this.context = this.$canvas.getContext('2d');
        this.resize(this.canvasConfig.width, this.canvasConfig.height);
    };
    Canvas2DContextService.prototype.getContext = function () {
        return this.context;
    };
    Canvas2DContextService.prototype.getDomElement = function () {
        return this.$canvas;
    };
    Canvas2DContextService.prototype.getDPR = function () {
        return this.dpr;
    };
    Canvas2DContextService.prototype.getBoundingClientRect = function () {
        if (this.$canvas.getBoundingClientRect) {
            return this.$canvas.getBoundingClientRect();
        }
    };
    Canvas2DContextService.prototype.destroy = function () {
        // @ts-ignore
        if (this.$container && this.$canvas && this.$canvas.parentNode) {
            // destroy context
            // @ts-ignore
            this.$container.removeChild(this.$canvas);
        }
    };
    Canvas2DContextService.prototype.resize = function (width, height) {
        var devicePixelRatio = this.canvasConfig.devicePixelRatio;
        // use user-defined dpr first
        var dpr = devicePixelRatio || (gLite.isBrowser && window.devicePixelRatio) || 1;
        dpr = dpr >= 1 ? Math.ceil(dpr) : 1;
        this.dpr = dpr;
        if (this.$canvas) {
            // set canvas width & height
            this.$canvas.width = this.dpr * width;
            this.$canvas.height = this.dpr * height;
            // set CSS style width & height
            gLite.setDOMSize(this.$canvas, width, height);
            // const dpr = this.getDPR();
            // scale all drawing operations by the dpr
            // @see https://www.html5rocks.com/en/tutorials/canvas/hidpi/
            // this.context.scale(dpr, dpr);
        }
        this.renderingContext.renderReasons.add(gLite.RenderReason.CAMERA_CHANGED);
    };
    Canvas2DContextService.prototype.applyCursorStyle = function (cursor) {
        if (this.$container && this.$container.style) {
            this.$container.style.cursor = cursor;
        }
    };
    Canvas2DContextService.prototype.toDataURL = function (options) {
        if (options === void 0) { options = {}; }
        return tslib.__awaiter(this, void 0, void 0, function () {
            var type, encoderOptions;
            return tslib.__generator(this, function (_a) {
                type = options.type, encoderOptions = options.encoderOptions;
                return [2 /*return*/, this.context.canvas.toDataURL(type, encoderOptions)];
            });
        });
    };
    return Canvas2DContextService;
}());

var ContextRegisterPlugin = /** @class */ (function (_super) {
    tslib.__extends(ContextRegisterPlugin, _super);
    function ContextRegisterPlugin() {
        var _this = _super.apply(this, tslib.__spreadArray([], tslib.__read(arguments), false)) || this;
        _this.name = 'canvas-context-register';
        return _this;
    }
    ContextRegisterPlugin.prototype.init = function () {
        this.context.ContextService = Canvas2DContextService;
    };
    ContextRegisterPlugin.prototype.destroy = function () {
        delete this.context.ContextService;
    };
    return ContextRegisterPlugin;
}(gLite.AbstractRendererPlugin));

var Renderer = /** @class */ (function (_super) {
    tslib.__extends(Renderer, _super);
    function Renderer(config) {
        var _this = _super.call(this, config) || this;
        // register Canvas2DContext
        _this.registerPlugin(new ContextRegisterPlugin());
        _this.registerPlugin(new ImageLoader__namespace.Plugin());
        _this.registerPlugin(new CanvasPathGenerator__namespace.Plugin());
        // enable rendering with Canvas2D API
        _this.registerPlugin(new CanvasRenderer__namespace.Plugin());
        _this.registerPlugin(new DomInteraction__namespace.Plugin());
        // enable picking with Canvas2D API
        _this.registerPlugin(new CanvasPicker__namespace.Plugin());
        // render HTML component
        _this.registerPlugin(new HTMLRenderer__namespace.Plugin());
        return _this;
    }
    return Renderer;
}(gLite.AbstractRenderer));

exports.CanvasPathGenerator = CanvasPathGenerator__namespace;
exports.CanvasPicker = CanvasPicker__namespace;
exports.CanvasRenderer = CanvasRenderer__namespace;
exports.DomInteraction = DomInteraction__namespace;
exports.HTMLRenderer = HTMLRenderer__namespace;
exports.ImageLoader = ImageLoader__namespace;
exports.Renderer = Renderer;
