// @antv/g6-extension-3d/esm/plugins/light.js
import { AmbientLight, DirectionalLight } from '@antv/g-plugin-3d';
import { BasePlugin, GraphEvent } from '@antv/g6';
import { deepMix } from '@antv/util';
export class Light extends BasePlugin {
    static defaultOptions = {
        ambient: {
            fill: '#fff',
            intensity: Math.PI * 2,
        },
        directional: {
            fill: '#fff',
            direction: [-1, 0, 1],
            intensity: Math.PI * 0.7,
        },
    };
    ambient;
    directional;
    constructor(context, options) {
        super(context, deepMix({}, Light.defaultOptions, options));
        this.bindEvents();
    }
    bindEvents() {
        this.context.graph.on(GraphEvent.BEFORE_DRAW, this.setLight);
    }
    unbindEvents() {
        this.context.graph.off(GraphEvent.BEFORE_DRAW, this.setLight);
    }
    setLight = () => {
        const { ambient, directional } = this.options;
        this.upsertLight('directional', directional);
        this.upsertLight('ambient', ambient);
    };
    upsertLight(type, options) {
        if (options) {
            const light = this[type];
            if (light)
                light.attr(options);
            else {
                const Ctor = type === 'ambient' ? AmbientLight : DirectionalLight;
                const light = new Ctor({ style: options });
                this[type] = light;
                this.context.canvas.appendChild(light);
            }
        }
        else
            this[type]?.remove();
    }
    destroy() {
        this.unbindEvents();
        super.destroy();
    }
}
