// @antv/g6-extension-3d/esm/elements/base-node-3d.js
import { Mesh } from '@antv/g-plugin-3d';
import { BaseNode, omitStyleProps, subStyleProps } from '@antv/g6';
import { deepMix } from '@antv/util';
import { createMaterial } from '../utils/material';
export class BaseNode3D extends BaseNode {
    static defaultStyleProps = {
        materialType: 'basic',
    };
    type = 'node-3d';
    get plugin() {
        const renderer = this.attributes.context.canvas.renderers['main'];
        const plugin = renderer.getPlugin('device-renderer');
        return plugin;
    }
    get device() {
        return this.plugin.getDevice();
    }
    constructor(options) {
        super(deepMix({}, { style: BaseNode3D.defaultStyleProps }, options));
    }
    render(attributes, container) {
        super.render(attributes, container);
    }
    getKeyStyle(attributes) {
        const style = omitStyleProps(super.getKeyStyle(attributes), 'material');
        const geometry = this.getGeometry(attributes);
        const material = this.getMaterial(attributes);
        return { x: 0, y: 0, z: 0, ...style, geometry, material };
    }
    drawKeyShape(attributes, container = this) {
        return this.upsert('key', Mesh, this.getKeyStyle(attributes), container);
    }
    getMaterial(attributes) {
        const { texture } = attributes;
        const materialStyle = subStyleProps(attributes, 'material');
        return createMaterial(this.plugin, materialStyle, texture);
    }
}
