// @antv/g6-extension-3d/lib/elements/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Torus = exports.Sphere = exports.Plane = exports.Line3D = exports.Cylinder = exports.Cube = exports.Cone = exports.Capsule = exports.BaseNode3D = void 0;
var base_node_3d_1 = require("./base-node-3d");
Object.defineProperty(exports, "BaseNode3D", { enumerable: true, get: function () { return base_node_3d_1.BaseNode3D; } });
var capsule_1 = require("./capsule");
Object.defineProperty(exports, "Capsule", { enumerable: true, get: function () { return capsule_1.Capsule; } });
var cone_1 = require("./cone");
Object.defineProperty(exports, "Cone", { enumerable: true, get: function () { return cone_1.Cone; } });
var cube_1 = require("./cube");
Object.defineProperty(exports, "Cube", { enumerable: true, get: function () { return cube_1.Cube; } });
var cylinder_1 = require("./cylinder");
Object.defineProperty(exports, "Cylinder", { enumerable: true, get: function () { return cylinder_1.Cylinder; } });
var line_3d_1 = require("./line-3d");
Object.defineProperty(exports, "Line3D", { enumerable: true, get: function () { return line_3d_1.Line3D; } });
var plane_1 = require("./plane");
Object.defineProperty(exports, "Plane", { enumerable: true, get: function () { return plane_1.Plane; } });
var sphere_1 = require("./sphere");
Object.defineProperty(exports, "Sphere", { enumerable: true, get: function () { return sphere_1.Sphere; } });
var torus_1 = require("./torus");
Object.defineProperty(exports, "Torus", { enumerable: true, get: function () { return torus_1.Torus; } });
