// @antv/g6-extension-3d/lib/behaviors/roll-canvas-3d.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RollCanvas3D = void 0;
const g6_1 = require("@antv/g6");
class RollCanvas3D extends g6_1.BasePlugin {
    static defaultOptions = {
        enable: true,
        trigger: ['wheel'],
        sensitivity: 1,
    };
    shortcut;
    get camera() {
        return this.context.canvas.getCamera();
    }
    constructor(context, options) {
        super(context, { ...RollCanvas3D.defaultOptions, ...options });
        this.shortcut = new g6_1.Shortcut(context.graph);
        this.bindEvents();
    }
    getAngle(delta) {
        const { sensitivity } = this.options;
        return -(delta * sensitivity) / 10;
    }
    onRoll = (event) => {
        const roll = this.camera.getRoll();
        const delta = event.deltaY;
        this.camera.setRoll(roll + this.getAngle(delta));
    };
    bindEvents() {
        const { trigger } = this.options;
        this.shortcut.unbindAll();
        this.shortcut.bind([...trigger, 'wheel'], this.onRoll);
    }
}
exports.RollCanvas3D = RollCanvas3D;
